File: python/1002.Find-Common-Characters.py
class Solution:
    def commonChars(self, words: List[str]) -> List[str]:
        curr_freq = Counter(words[0])

        for i in range(1, len(words)):
            temp_freq = Counter(words[i])

            for key in list(curr_freq.keys()):
                val = temp_freq.get(key, 0)

                if val == 0:
                    del curr_freq[key]
                else:
                    curr_freq[key] = min(curr_freq[key], val)

        result = []

        for key, val in curr_freq.items():
            result.extend([key] * val)

        return result

File: javascript/1002.Find-Common-Characters.js
/**
 * @param {string[]} words
 * @return {string[]}
 */
var commonChars = function (words) {
  let result = "";

  const createMap = (word) => {
    const map = new Map();

    for (const c of word) {
      map.set(c, (map.get(c) || 0) + 1);
    }

    return map;
  };

  const currMap = createMap(words[0]);

  for (let i = 1; i < words.length; i++) {
    const tempMap = createMap(words[i]);

    for (const key of [...currMap.keys()]) {
      const val = tempMap.get(key) || 0;

      if (val === 0) {
        currMap.delete(key);
      } else {
        currMap.set(key, Math.min(currMap.get(key), val));
      }
    }
  }

  for (const [key, val] of currMap) {
    result += key.repeat(val);
  }

  return result.split("");
};
